frame/uniswap-v2/src/tests.rs

```diff
@@ -0,0 +1,148 @@
+		assert_ok!(Tokens::mint_into(MockCurrencyId::USDT, &BOB, 200000));
+		assert_eq!(Tokens::balance(MockCurrencyId::USDT, &BOB), 200000);
+		assert_eq!(Tokens::balance(MockCurrencyId::USDC, &BOB), 0);
+		assert_ok!(Tokens::mint_into(MockCurrencyId::USDC, &BOB, 200000));
+		assert_eq!(Tokens::balance(MockCurrencyId::USDC, &BOB), 200000);
+		let p = ConstantProductAmm::create_pool(&ALICE, assets, amp_coeff, fee, admin_fee);
+		assert_ok!(&p);
+		let pool_id = p.unwrap();
+		let pool = ConstantProductAmm::get_pool_info(pool_id);
+		assert!(pool.is_some());
+		let pool_lp_asset = ConstantProductAmm::pool_lp_asset(pool_id);
+		assert!(pool_lp_asset.is_some());
+		let pool_lp_asset = pool_lp_asset.unwrap();
+		// 1 USDC = 1 USDT
+		let amounts = vec![130000u128, 130000u128];
+		assert_ok!(ConstantProductAmm::add_liquidity(&ALICE, pool_id, amounts.clone(), 0u128));
+		let alice_balance = Tokens::balance(pool_lp_asset, &ALICE);
+		assert_ne!(alice_balance, 0);
+		assert_eq!(Tokens::balance(MockCurrencyId::USDT, &ALICE), 200000 - 130000);
+		assert_eq!(Tokens::balance(MockCurrencyId::USDC, &ALICE), 200000 - 130000);
+		let pool = ConstantProductAmm::get_pool_info(pool_id);
+		assert!(pool.is_some());
+		assert_ok!(ConstantProductAmm::add_liquidity(&BOB, pool_id, amounts.clone(), 0u128));
+		let bob_balance = Tokens::balance(pool_lp_asset, &BOB);
+		assert_ne!(bob_balance, 0);
+		assert_eq!(Tokens::balance(MockCurrencyId::USDT, &BOB), 200000 - 130000);
+		assert_eq!(Tokens::balance(MockCurrencyId::USDC, &BOB), 200000 - 130000);
+		let min_amt = vec![0u128, 0u128];
+		assert_eq!(
+			Tokens::balance(MockCurrencyId::USDC, &ConstantProductAmm::account_id(&pool_id)),
+			260000
+		);
+		assert_eq!(
+			Tokens::balance(MockCurrencyId::USDT, &ConstantProductAmm::account_id(&pool_id)),
+			260000
+		);
+		assert_ok!(ConstantProductAmm::remove_liquidity(
+			&ALICE,
+			pool_id,
+			alice_balance,
+			min_amt.clone()
+		));
+		assert_eq!(Tokens::balance(pool_lp_asset, &ALICE), 0);
+		assert_eq!(Tokens::balance(MockCurrencyId::USDT, &ALICE), 200000);
+		assert_eq!(Tokens::balance(MockCurrencyId::USDC, &ALICE), 200000);
+		assert_eq!(
+			Tokens::balance(MockCurrencyId::USDC, &ConstantProductAmm::account_id(&pool_id)),
+			130000
+		);
+		assert_eq!(
+			Tokens::balance(MockCurrencyId::USDT, &ConstantProductAmm::account_id(&pool_id)),
+			130000
+		);
+		assert_ok!(ConstantProductAmm::remove_liquidity(
+			&BOB,
+			pool_id,
+			bob_balance,
+			min_amt.clone()
+		));
+		assert_eq!(Tokens::balance(pool_lp_asset, &BOB), 0);
+		assert_eq!(Tokens::balance(MockCurrencyId::USDT, &BOB), 200000);
+		assert_eq!(Tokens::balance(MockCurrencyId::USDC, &BOB), 200000);
+		assert_eq!(
+			Tokens::balance(MockCurrencyId::USDC, &ConstantProductAmm::account_id(&pool_id)),
+			0
+		);
+		assert_eq!(
+			Tokens::balance(MockCurrencyId::USDT, &ConstantProductAmm::account_id(&pool_id)),
+			0
+		);
+	});
+}
+
+#[test]
+fn exchange_test() {
```

Comment:
The tests are too big to be understandable. We should cut this kind of test or comment some part to make it more readable. I personnally think we should cut this into a lot of scenarios.

Reply:
Added comments